hathor/feature_activation/feature_service.py

```diff
@@ -0,0 +1,170 @@
+        """Returns the new feature state based on the new block, the criteria, and the previous state."""
+        assert boundary_block.height % self._settings.evaluation_interval == 0, (
+            'cannot calculate new state for a non-boundary block'
+        )
+        criteria = self._get_criteria(feature=feature)
+
+        if previous_state is FeatureState.DEFINED:
+            if boundary_block.height >= criteria.start_height:
+                return FeatureState.STARTED
+
+            return FeatureState.DEFINED
+
+        if previous_state is FeatureState.STARTED:
+            if boundary_block.height >= criteria.timeout_height and not criteria.activate_on_timeout:
+                return FeatureState.FAILED
+
+            if (
+                boundary_block.height >= criteria.timeout_height
+                and criteria.activate_on_timeout
+                and boundary_block.height >= criteria.minimum_activation_height
+            ):
+                return FeatureState.ACTIVE
+
+            count = self.get_bit_count(boundary_block=boundary_block.data, bit=criteria.bit)
+            threshold = criteria.threshold if criteria.threshold is not None else self._settings.default_threshold
+
+            if (
+                boundary_block.height < criteria.timeout_height
+                and count >= threshold
+                and boundary_block.height >= criteria.minimum_activation_height
+            ):
+                return FeatureState.ACTIVE
+
+            return FeatureState.STARTED
+
+        if previous_state is FeatureState.ACTIVE:
+            return FeatureState.ACTIVE
+
+        if previous_state is FeatureState.FAILED:
+            return FeatureState.FAILED
+
+        raise ValueError(f'Unknown previous state: {previous_state}')
+
+    def _get_criteria(self, *, feature: Feature) -> Criteria:
+        criteria = self._settings.features.get(feature)
+
+        if not criteria:
+            raise ValueError(f"Criteria not defined for feature '{feature}'.")
+
+        return criteria
+
+    def get_bits_description(self, *, block: Block) -> dict[Feature, tuple[Criteria, FeatureState]]:
+        """Returns the criteria definition and feature state for all features at a certain block."""
+        return {
+            feature: (criteria, self.get_state(block=block, feature=feature))
+            for feature, criteria in self._settings.features.items()
+        }
+
+    def get_bit_count(self, *, boundary_block: Block, bit: int) -> int:
+        """Returns the count of blocks with this bit enabled in the previous evaluation interval."""
+        assert not boundary_block.is_genesis, 'cannot calculate bit count for genesis'
+        assert boundary_block.calculate_height() % self._settings.evaluation_interval == 0, (
+            'cannot calculate bit count for a non-boundary block'
+        )
+        count = 0
+        block = boundary_block
+
+        for _ in range(self._settings.evaluation_interval):
```

Comment:
Should we have a new metadata that carries this sum? Something like the following:

Let `h` be the height, then:

```
count[h] = count[h-1]
if block[h].is_bit_active:
    count[h] += 1
if block[h - evaluation interval].is_bit_active:
    count[h] -= 1
```

If would allow us to access it in `O(1)` instead of `O(evaluation_interval)`.

Reply:
That sounds like a great idea. May it be implemented in a separate PR, as an improvement? Just to keep this one lean. I'll start working on it right away, though

Reply:
Ok. Just leave a note saying that this implementation might be improved using a metadata, please.

Reply:
Added the note in https://github.com/HathorNetwork/hathor-core/commit/537e883413c022a020973b99b20ce2e08874460a

Reply:
The PR with the actual implementation is almost ready

Reply:
Great! Thanks!

Reply:
Implemented in https://github.com/HathorNetwork/hathor-core/pull/633